app/models/concerns/spree/gift_cards/order_concerns.rb

```diff
@@ -11,17 +11,22 @@ module GiftCards::OrderConcerns
     end
 
     module InstanceMethods
-      def finalize!
-        create_gift_cards
-        super
+      def gift_card_match(line_item, options)
+        gift_card_options = options["gift_card_details"]
+        !line_item.gift_card? ||
+          (line_item.gift_card? && gift_card_options.present? &&
+          line_item.gift_cards.any? {|gc| gc.recipient_email == gift_card_options['recipient_email']} &&
+          line_item.gift_cards.any? {|gc| gc.recipient_name == gift_card_options['recipient_name']} &&
+          line_item.gift_cards.any? {|gc| gc.purchaser_name == gift_card_options['purchaser_name']} &&
+          line_item.gift_cards.any? {|gc| gc.gift_message == gift_card_options['gift_message']})
       end
 
-      def create_gift_cards
-        line_items.each do |item|
-          item.quantity.times do
-            Spree::VirtualGiftCard.create!(amount: item.price, currency: item.currency, purchaser: user, line_item: item) if item.gift_card?
-          end
+      def finalize!
+        gift_cards.each do |gift_card|
```

Comment:
what happens if one of the later order finalization steps fails? seems to me this should go after the super behavior, not before